modules/metrics/src/main/java/org/apache/ignite/internal/metrics/composite/DistributionMetric.java

```diff
@@ -0,0 +1,144 @@
+import org.apache.ignite.internal.metrics.AbstractMetric;
+import org.apache.ignite.internal.metrics.Metric;
+import org.apache.ignite.internal.metrics.scalar.LongGauge;
+import org.apache.ignite.internal.metrics.scalar.LongMetric;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Distribution metric calculates counts of measurements that gets into each bounds interval.
+ * Note, that {@link #value()} will return array length of {@code bounds.length + 1}.
+ * Last element contains count of measurements bigger than most right value of bounds.
+ */
+public class DistributionMetric extends AbstractMetric implements CompositeMetric {
+    /** Count of measurement for each bound. */
+    private final AtomicLongArray measurements;
+
+    /** Bounds of measurements. */
+    private final long[] bounds;
+
+    /**
+     * The constructor.
+     *
+     * @param name Name.
+     * @param desc Description.
+     * @param bounds Bounds of the buckets.
+     */
+    public DistributionMetric(String name, @Nullable String desc, long[] bounds) {
+        super(name, desc);
+
+        assert bounds != null && bounds.length > 0;
+        assert isSorted(bounds);
+
+        this.bounds = bounds;
+        this.measurements = new AtomicLongArray(bounds.length + 1);
+    }
+
+    /**
+     * Check whether given array is sorted.
+     *
+     * @param arr Array to check.
+     * @return {@code True} if array sorted, {@code false} otherwise.
+     */
+    public static boolean isSorted(long[] arr) {
+        if (arr == null || arr.length < 2) {
+            return true;
+        }
+
+        for (int i = 1; i < arr.length; i++) {
+            if (arr[i - 1] > arr[i]) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    /**
+     * Sets value.
+     *
+     * @param x Value.
+     */
+    public void value(long x) {
+        assert x >= 0;
+
+        //Expect arrays of few elements.
+        for (int i = 0; i < bounds.length; i++) {
+            if (x <= bounds[i]) {
+                measurements.incrementAndGet(i);
+
+                return;
+            }
+        }
+
+        measurements.incrementAndGet(bounds.length);
+    }
+
+    /** {@inheritDoc} */
+    public long[] value() {
+        long[] res = new long[measurements.length()];
+
+        for (int i = 0; i < measurements.length(); i++) {
+            res[i] = measurements.get(i);
+        }
+
+        return res;
+    }
+
+    /** {@inheritDoc} */
+    @Override public @Nullable String getAsString() {
+        return asScalarMetrics().stream()
+            .map(m -> m.name() + ":" + ((LongMetric) m).value())
+            .collect(Collectors.joining(", ", "[", "]"));
+    }
+
+    /**
+     * Bounds of the buckets of distribution.
+     *
+     * @return Bounds of the buckets of distribution.
+     */
+    public long[] bounds() {
+        return bounds;
+    }
+
+    /** {@inheritDoc} */
+    @Override public List<Metric> asScalarMetrics() {
```

Comment:
A lot of actions. You can do it only once and always return this list just because you use gauges that always will return actual values from measurements.
 
Also, return list should be immutable.